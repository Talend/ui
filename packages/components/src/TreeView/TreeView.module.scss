@use '~@talend/bootstrap-theme/src/theme/guidelines' as *;
@use '~@talend/design-tokens/lib/tokens';

// TreeView styles

.tc-treeview {
	.tc-treeview-header {
		display: flex;
		justify-content: space-between;
		align-items: center;

		color: tokens.$coral-color-accent-text;
		line-height: 6rem;
		font-weight: bold;

		padding-left: $padding-large;
		border-bottom: 1px solid tokens.$coral-color-neutral-border;

		:global(.btn-link .tc-svg-icon) {
			stroke: tokens.$coral-color-accent-text;
		}
	}

	.tc-treeview-list {
		padding: $padding-small 0;
		overflow: auto;
		list-style-type: none;
	}
}
