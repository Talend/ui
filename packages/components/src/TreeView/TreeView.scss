// TreeView styles

.tc-treeview {
	.tc-treeview-header {
		display: flex;
		justify-content: space-between;
		align-items: center;

		color: $brand-primary;
		line-height: 6rem;
		font-weight: bold;
		text-transform: uppercase;

		padding-left: $padding-large;
		border-bottom: 1px rgba(0, 0, 0, 0.12) solid;

		:global(.btn-link .tc-svg-icon) {
			stroke: $brand-primary;
		}
	}

	.tc-treeview-list {
		padding: $padding-small 0;
		overflow: auto;
		list-style-type: none;
	}
}
