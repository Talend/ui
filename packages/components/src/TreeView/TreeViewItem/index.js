import TreeViewItem from './TreeViewItem.component';

export default TreeViewItem;
