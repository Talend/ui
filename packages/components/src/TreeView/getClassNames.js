import classNames from 'classnames';

const getClassName = theme => className => classNames(className, theme[className]);
export default getClassName;
