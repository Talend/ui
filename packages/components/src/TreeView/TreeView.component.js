import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import { Action } from '../';
import TreeViewItem from './TreeViewItem/';

import theme from './TreeView.scss';
import withTreeGesture from '../Tree/withTreeGesture';

/**
 * A view component to display any tree structure, like folders or categories.
 *
 * @param id, for qa purposes
 * @param structure optional, tree structure to display, see example below
 * @param headerText optional, specifies text in component's header
 * @param addAction optional, defines if 'add' button is displayed and
 *        specifies button click event callback
 * @param addActionLabel optional, specifies tooltip label for 'add' button
 * @param itemSelectCallback required, tree item click event callback function
 * @param itemToggleCallback required, tree item expand/collapse event callback function
 *
 * const defaultProps = {
 * 	structure: [{
 * 		name: 'grandpa',
 * 		children: [
 * 			{
 * 				name: 'mami',
 * 				toggled: true,
 * 				children: [
 * 					{ name: 'me', selected: true },
 * 					{ name: 'bro' },
 * 				],
 * 			},
 * 			{ name: 'aunt', toggled: false, children: [{ name: 'cousin' }] },
 * 		],
 * 		toggled: true
 * 	}],
 * 	headerText: 'some elements',
 * 	addAction: () => null,
 * 	addActionLabel: 'add element',
 * 	itemSelectCallback: () => null,
 * 	itemToggleCallback: () => null,
 * }
 *
 * <TreeView {...defaultProps} />
 *
 */
function TreeView(props) {
	const {
		id,
		headerText,
		structure,
		addAction,
		addActionLabel,
		onKeyDown,
		onSelect,
		onToggle,
		noHeader,
		className,
		selectedId,
		style,
	} = props;
	const titleId = id && `${id}-title`;
	return (
		<div className={classNames('tc-treeview', theme['tc-treeview'], className)} style={style}>
			<header className={classNames(theme['tc-treeview-header'], { 'sr-only': noHeader })}>
				<span id={titleId}>{headerText}</span>
				{addAction && (
					<Action
						label={addActionLabel}
						icon="talend-plus"
						onClick={addAction}
						tooltipPlacement="right"
						hideLabel
						link
						id={id && `${id}-add`}
						key={addActionLabel}
					/>
				)}
			</header>
			<ul className={theme['tc-treeview-list']} role="tree" aria-labelledby={titleId}>
				{structure.map((item, i) => (
					<TreeViewItem
						id={id && `${id}-${i}`}
						item={item}
						siblings={structure}
						onKeyDown={onKeyDown}
						onSelect={onSelect}
						onToggle={onToggle}
						key={i}
						index={i + 1}
						selectedId={selectedId}
						level={1}
					/>
				))}
			</ul>
		</div>
	);
}

TreeView.displayName = 'TreeView';

TreeView.defaultProps = {
	id: 'tc-treeview',
	addActionLabel: 'Add folder',
	headerText: 'Folders',
};

if (process.env.NODE_ENV !== 'production') {
	TreeView.propTypes = {
		id: PropTypes.string.isRequired,
		headerText: PropTypes.string,
		structure: PropTypes.arrayOf(TreeViewItem.propTypes.item),
		addAction: PropTypes.func,
		addActionLabel: PropTypes.string,
		onKeyDown: PropTypes.func.isRequired,
		onToggle: PropTypes.func.isRequired,
		onSelect: PropTypes.func.isRequired,
		noHeader: PropTypes.bool,
		className: PropTypes.string,
		selectedId: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
		style: PropTypes.object,
	};
}

export default withTreeGesture(TreeView);
