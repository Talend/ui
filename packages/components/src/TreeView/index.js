import TreeView from './TreeView.component';

export default TreeView;
