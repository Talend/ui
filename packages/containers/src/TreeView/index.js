import TreeView from './TreeView.container';

export default TreeView;
