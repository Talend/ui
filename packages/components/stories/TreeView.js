/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { checkA11y } from '@storybook/addon-a11y';
import { action } from '@storybook/addon-actions';

import { TreeView, IconsProvider } from '../src/index';

const structure = [
	{ name: 'hitmonlee', children: [{ name: 'Hitmonchan' }], isOpened: false },
	{ name: 'pikachu', children: [{ name: 'raichu' }], isOpened: true },
	{
		id: 'selected',
		name: 'Abra',
		isOpened: true,
		children: [
			{
				name: 'Kadabra',
				isOpened: true,
				children: [
					{
						name: 'Alakazam',
					},
				],
			},
		],
	},
];

const structureWithIcons = [
	{
		name: 'hitmonlee',
		children: [{ name: 'Hitmonchan' }],
		isOpened: false,
		icon: {
			name: 'src-http://static.pokemonpets.com/images/monsters-images-300-300/106-Hitmonlee.png',
		},
	},
	{
		name: 'pikachu',
		children: [
			{
				name: 'raichu',
				icon: {
					name:
						'src-http://static.pokemonpets.com/images/monsters-images-300-300/2026-Shiny-Raichu.png',
				},
			},
		],
		isOpened: true,
		icon: {
			name:
				'src-http://static.pokemonpets.com/images/monsters-images-300-300/2025-Shiny-Pikachu.png',
		},
	},
	{
		id: 'selected',
		name: 'Abra',
		icon: { name: 'src-http://static.pokemonpets.com/images/monsters-images-300-300/63-Abra.png' },
		isOpened: true,
		children: [
			{
				name: 'Kadabra',
				icon: {
					name: 'src-http://static.pokemonpets.com/images/monsters-images-300-300/64-Kadabra.png',
				},
				isOpened: true,
				children: [
					{
						name: 'Alakazam',
						icon: {
							name:
								'src-http://static.pokemonpets.com/images/monsters-images-300-300/65-Alakazam.png',
						},
					},
				],
			},
		],
	},
];

const removeAction = [
	{
		action: action('itemRemoveCallback'),
		icon: 'talend-trash',
		label: 'remove element',
	},
];

const actions = [
	{
		action: action('itemAddCallback'),
		icon: 'talend-plus',
		label: 'Add Item',
	},
	{
		action: action('itemEditCallback'),
		icon: 'talend-pencil',
		label: 'Edit Item',
	},
	{
		action: action('itemRemoveCallback'),
		icon: 'talend-trash',
		label: 'Remove Item',
	},
];

const structureWithCounterAndAction = [
	{
		name: 'hitmonlee',
		isOpened: true,
		children: [
			{
				name: 'raichu',
				showCounter: true,
				counter: 111,
				actions: removeAction,
			},
		],
		counter: -1,
		showCounter: true,
		actions: removeAction,
	},
	{
		name: 'pikachu',
		isOpened: true,
		counter: 2911,
		showCounter: true,
		actions: removeAction,
	},
];

const structureWithActions = [
	{
		name: 'hitmonlee',
		isOpened: true,
		children: [
			{
				name: 'raichu',
				actions,
			},
		],
		actions,
	},
	{
		name: 'pikachu',
		isOpened: true,
		actions,
	},
];

const defaultProps = {
	id: 'my-treeview',
	structure,
	onSelect: action('onSelect'),
	onToggle: action('onToggle'),
	onToggleAllSiblings: action('onToggleAllSiblings'),
	selectedId: 'selected',
};

const withAddAction = {
	...defaultProps,
	addAction: action('added'),
};

const withHeader = {
	...withAddAction,
	headerText: 'some elements',
	addActionLabel: 'add element',
};

const withRemoval = {
	...withAddAction,
};
withRemoval.structure = structureWithCounterAndAction;

const withActions = {
	...withAddAction,
};
withActions.structure = structureWithActions;

const hugeStructure = [
	{
		name: 'Hitmonlee1',
		isOpened: true,
		children: [
			{
				name: 'Hitmonchan2',
				isOpened: true,
				children: [
					{
						name: 'Hitmonchan3',
						isOpened: true,
						children: [
							{
								name: 'Hitmonchan4',
								isOpened: true,
								children: [
									{
										name: 'Hitmonchan5',
										isOpened: true,
										children: [
											{
												name: 'Hitmonchan6',
												isOpened: true,
												children: [
													{
														name: 'Hitmonchan7',
														isOpened: true,
														children: [
															{
																name: 'Hitmonchan8',
																isOpened: true,
																children: [
																	{
																		name: 'Hitmonchan9',
																		isOpened: true,
																		children: [
																			{
																				name: 'Hitmonchan10',
																				isOpened: true,
																				children: [
																					{
																						name: 'Hitmonchen11',
																						isOpened: true,
																						children: [
																							{
																								name: 'Hitmonchen12',
																								isOpened: true,
																								children: [
																									{
																										name: 'Hitmonchen13',
																										isOpened: true,
																										children: [
																											{
																												name: 'Hitmonchen14',
																											},
																										],
																									},
																								],
																							},
																						],
																					},
																				],
																			},
																		],
																	},
																],
															},
														],
													},
												],
											},
										],
									},
								],
							},
						],
					},
				],
			},
		],
	},
];

const withDeepStructure = {
	...defaultProps,
};
withDeepStructure.structure = hugeStructure;

const cornerCaseLongName = {
	...defaultProps,
};
cornerCaseLongName.structure = [
	{
		name:
			'Hitmonlee1Hitmonlee1Hitmonlee1Hitmonlee1Hitmonlee1 Hitmonlee1Hitmonlee1Hitmonlee1Hitmonlee1Hitmonlee1',
		isOpened: true,
	},
];

const style = { width: '300px', border: '1px solid #eee', marginLeft: '10px' };

storiesOf('TreeView', module)
	.addDecorator(checkA11y)
	.addWithInfo('default', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set with action example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withAddAction} />
			</div>
		</div>
	))
	.addWithInfo('with custom icons', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>The icons can be customized, passign the Icon components props</p>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withAddAction} structure={structureWithIcons} />
			</div>
		</div>
	))
	.addWithInfo('with custom header text', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Custom header and action tooltip property-set example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withHeader} />
			</div>
		</div>
	))
	.addWithInfo('without action', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set without action example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...defaultProps} />
			</div>
		</div>
	))
	.addWithInfo('without header', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set without header example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...defaultProps} noHeader />
			</div>
		</div>
	))
	.addWithInfo('with remove action and counter', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set with remove action example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withRemoval} />
			</div>
		</div>
	))
	.addWithInfo('with many actions', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set with remove action example: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withActions} />
			</div>
		</div>
	))
	.addWithInfo('with deep structure', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set with deep structure: </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...withDeepStructure} />
			</div>
		</div>
	))
	.addWithInfo('with long name', () => (
		<div>
			<h1>TreeView</h1>
			<h3>Definition</h3>
			<p>A view component to display any tree structure, like folders or categories.</p>
			<h3>Default property-set with cornercase: longname </h3>
			<div style={style}>
				<IconsProvider />
				<TreeView {...cornerCaseLongName} />
			</div>
		</div>
	));
