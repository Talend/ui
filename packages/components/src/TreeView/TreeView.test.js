import React from 'react';
import { shallow } from 'enzyme';

import TreeView from './TreeView.component';

const defaultProps = {
	id: 'id',
	structure: [
		{
			name: 'grandpa',
			actions: [
				{
					action: () => 'itemRemoveCallback',
					icon: 'talend-trash',
					label: 'remove element',
				},
			],
			children: [
				{
					name: 'mami',
					toggled: true,
					children: [{ name: 'me', selected: true }, { name: 'bro' }],
				},
				{ name: 'aunt', toggled: false, children: [{ name: 'cousin' }] },
			],
			toggled: true,
			counter: 101,
			showCounter: true,
		},
	],
	headerText: 'some elements',
	addAction: jest.fn(),
	addActionLabel: 'add element',
	onSelect: jest.fn(),
	onToggle: jest.fn(),
	onToggleAllSiblings: jest.fn(),
};

describe('TreeView', () => {
	it('should render', () => {
		// when
		const wrapper = shallow(<TreeView {...defaultProps} />);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('when a user click on the add Action it should call props.addAction', () => {
		// given
		const props = {
			...defaultProps,
			addAction: jest.fn(),
		};
		const wrapper = shallow(<TreeView {...props} />);
		expect(props.addAction).not.toBeCalled();

		// when
		wrapper.find('Action').simulate('click');

		// then
		expect(props.addAction).toBeCalled();
	});
});
