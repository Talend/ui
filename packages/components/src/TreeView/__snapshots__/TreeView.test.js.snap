// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView should render 1`] = `
<div
  className="tc-treeview theme-tc-treeview"
  style={undefined}
>
  <header
    className="theme-tc-treeview-header"
  >
    <span
      id="id-title"
    >
      some elements
    </span>
    <Action
      hideLabel={true}
      icon="talend-plus"
      id="id-add"
      label="add element"
      link={true}
      onClick={[Function]}
      tooltipPlacement="right"
    />
  </header>
  <ul
    aria-labelledby="id-title"
    className="theme-tc-treeview-list"
    role="tree"
  >
    <TreeViewItem
      id="id-0"
      index={1}
      item={
        Object {
          "actions": Array [
            Object {
              "action": [Function],
              "icon": "talend-trash",
              "label": "remove element",
            },
          ],
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "name": "me",
                  "selected": true,
                },
                Object {
                  "name": "bro",
                },
              ],
              "name": "mami",
              "toggled": true,
            },
            Object {
              "children": Array [
                Object {
                  "name": "cousin",
                },
              ],
              "name": "aunt",
              "toggled": false,
            },
          ],
          "counter": 101,
          "name": "grandpa",
          "showCounter": true,
          "toggled": true,
        }
      }
      itemSiblings={
        Array [
          Object {
            "actions": Array [
              Object {
                "action": [Function],
                "icon": "talend-trash",
                "label": "remove element",
              },
            ],
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "name": "me",
                    "selected": true,
                  },
                  Object {
                    "name": "bro",
                  },
                ],
                "name": "mami",
                "toggled": true,
              },
              Object {
                "children": Array [
                  Object {
                    "name": "cousin",
                  },
                ],
                "name": "aunt",
                "toggled": false,
              },
            ],
            "counter": 101,
            "name": "grandpa",
            "showCounter": true,
            "toggled": true,
          },
        ]
      }
      level={1}
      onSelect={[Function]}
      onToggle={[Function]}
      onToggleAllSiblings={[Function]}
      selectedId={undefined}
    />
  </ul>
</div>
`;
